tier1/keyvalues3.cpp

```diff
@@ -0,0 +1,1607 @@
+	}
+
+	m_IsExternalName.AddToTail( false );
+
+	if ( m_pFastSearch && !m_pFastSearch->m_ignore )
+		m_pFastSearch->m_member_ids.Insert( name.GetHashCode(), memberId );
+
+	return memberId;
+}
+
+void CKeyValues3Table::CopyFrom( const CKeyValues3Table* pSrc )
+{
+	int nNewSize = pSrc->m_Hashes.Count();
+
+	RemoveAll( nNewSize );
+
+	m_Hashes.SetCount( nNewSize );
+	m_Members.SetCount( nNewSize );
+	m_Names.SetCount( nNewSize );
+	m_IsExternalName.SetCount( nNewSize );
+
+	CKeyValues3Context* context = GetContext();
+
+	for ( int i = 0; i < nNewSize; ++i )
+	{
+		m_Hashes[i] = pSrc->m_Hashes[i];
+
+		if ( context )
+		{
+			m_Members[i] = context->AllocKV();
+			m_Names[i] = context->AllocString( pSrc->m_Names[i] );
+		}
+		else
+		{
+			m_Members[i] = new KeyValues3;
+			m_Names[i] = strdup( pSrc->m_Names[i] );
+		}
+
+		m_IsExternalName[i] = false;
+
+		*m_Members[i] = *pSrc->m_Members[i];
+	}
+
+	if ( nNewSize >= 128 )
+		EnableFastSearch();
+}
+
+void CKeyValues3Table::RemoveMember( KV3MemberId_t id )
+{
+	CKeyValues3Context* context = GetContext();
+
+	if ( context ) 
+	{
+		context->FreeKV( m_Members[ id ] );
+	}
+	else
+	{
+		delete m_Members[ id ];
+		delete m_Names[ id ];
+	}
+
+	m_Hashes.Remove( id );
+	m_Members.Remove( id );
+	m_Names.Remove( id );
+	m_IsExternalName.Remove( id );
+
+	if ( m_pFastSearch )
+	{
+		m_pFastSearch->m_ignore = true;
+		m_pFastSearch->m_ignores_counter = 1;
+	}
+}
+
+void CKeyValues3Table::RemoveAll( int nAllocSize )
+{
+	CKeyValues3Context* context = GetContext();
+
+	for ( int i = 0; i < m_Hashes.Count(); ++i )
+	{
+		if ( context )
+		{
+			context->FreeKV( m_Members[i] );
+		}
+		else
+		{
+			delete m_Members[i]; 
+			delete m_Names[i];
+		}
+	}
+
+	m_Hashes.RemoveAll();
+	m_Members.RemoveAll();
+	m_Names.RemoveAll();
+	m_IsExternalName.RemoveAll();
+
+	if ( nAllocSize > 0 )
+	{
+		m_Hashes.EnsureCapacity( nAllocSize );
+		m_Members.EnsureCapacity( nAllocSize );
+		m_Names.EnsureCapacity( nAllocSize );
+		m_IsExternalName.EnsureCapacity( nAllocSize );
+	}
+
+	if ( m_pFastSearch )
+	{
+		if ( nAllocSize >= 128 )
+		{
+			m_pFastSearch->Clear();
+		}
+		else
+		{
+			delete m_pFastSearch;
+			m_pFastSearch = NULL;
+		}
+	}
+}
+
+void CKeyValues3Table::Purge( bool bClearingContext )
+{
+	CKeyValues3Context* context = GetContext();
+
+	for ( int i = 0; i < m_Hashes.Count(); ++i )
+	{
+		if ( context )
+		{
+			if ( !bClearingContext )
+				context->FreeKV( m_Members[i] );
+		}
+		else
+		{
+			delete m_Members[i]; 
+			delete m_Names[i];
+		}
+	}
+
+	if ( m_pFastSearch )
+		delete m_pFastSearch;
+	m_pFastSearch = NULL;
+
+	m_Hashes.Purge();
+	m_Members.Purge();
+	m_Names.Purge();
+	m_IsExternalName.Purge();
+}
+
+KeyValues3* CKeyValues3Cluster::Alloc( KV3TypeEx_t type, KV3SubType_t subtype )
+{
+	int element = BitScanFwd( ~m_nAllocatedElements );
+	m_nAllocatedElements |= ( 1ull << element );
+	KeyValues3* kv = &m_KeyValues[ element ];
+	new( kv ) KeyValues3( element, type, subtype );
```

Comment:
This line conflicts with ``memdbgon.h`` and causes a compile time error when compiled in debug